rivermigrate/river_migrate.go

```diff
@@ -0,0 +1,502 @@
+//	if err != nil {
+//		// handle error
+//	}
+//	defer dbPool.Close()
+//
+//	migrator, err := rivermigrate.New(riverpgxv5.New(dbPool), nil)
+//	if err != nil {
+//		// handle error
+//	}
+func New[TTx any](driver riverdriver.Driver[TTx], config *Config) *Migrator[TTx] {
+	if config == nil {
+		config = &Config{}
+	}
+
+	logger := config.Logger
+	if logger == nil {
+		logger = slog.New(slog.NewTextHandler(os.Stdout, &slog.HandlerOptions{
+			Level: slog.LevelWarn,
+		}))
+	}
+
+	archetype := &baseservice.Archetype{
+		Logger:     logger,
+		TimeNowUTC: func() time.Time { return time.Now().UTC() },
+	}
+
+	return baseservice.Init(archetype, &Migrator[TTx]{
+		driver:     driver,
+		migrations: riverMigrationsMap,
+		queries:    dbsqlc.New(),
+	})
+}
+
+// MigrateOpts are options for a migrate operation.
+type MigrateOpts struct {
+	// MaxSteps is the maximum number of migrations to apply either up or down.
+	// Leave zero for an unlimited number. Set to -1 to apply no migrations (for
+	// testing/checking purposes).
+	MaxSteps int
+
+	// TargetVersion is a specific migration version to apply migrations to. The
+	// version must exist and it must be in the possible list of migrations to
+	// apply. e.g. If requesting an up migration with version 3, version 3 not
+	// already be applied.
+	//
+	// When applying migrations up, migrations are applied including the target
+	// version, so when starting at version 0 and requesting version 3, versions
+	// 1, 2, and 3 would be applied. When applying migrations down, down
+	// migrations are applied excluding the target version, so when starting at
+	// version 5 an requesting version 3, down migrations for versions 5 and 4
+	// would be applied, leaving the final schema at version 3.
+	TargetVersion int
+}
+
+// MigrateResult is the result of a migrate operation.
+type MigrateResult struct {
+	// Versions are migration versions that were added (for up migrations) or
+	// removed (for down migrations) for this run.
+	Versions []MigrateVersion
+}
+
+// MigrateVersion is the result for a single applied migration.
+type MigrateVersion struct {
+	// Version is the version of the migration applied.
+	Version int
+}
+
+func migrateVersionToInt(version MigrateVersion) int     { return version.Version }
+func migrateVersionToInt64(version MigrateVersion) int64 { return int64(version.Version) }
+
+type Direction string
+
+const (
+	DirectionDown Direction = "down"
+	DirectionUp   Direction = "up"
+)
+
+// Migrate migrates the database in the given direction (up or down). The opts
+// parameter may be omitted for convenience.
+//
+// By default, applies all outstanding migrations possible in either direction.
+// When migrating up all outstanding migrations are applied, and when migrating
+// down all existing migrations are unapplied.
+//
+// When migrating down, use with caution. MigrateOpts.MaxSteps should be set to
+// 1 to only migrate down one step.
```

Comment:
This seems like a bit of a footgun. I would have been inclined to default to running a single migration at a time just to avoid this. Do you think it's too risky?

At the very least we might want the examples to show the single-migration usage.

Reply:
I could see running one at a time to be somewhat defensible, but my fear is that it'll be the wrong thing for what might end up being the most common use of the Go migration API, which is to raise tables for test and development purposes.

In these cases what you most want out of the migrate API is for it to give you a fully functional River schema. Running one migration at a time doesn't do that, so you'd either have to hammer it into place by setting `MaxSteps: 1000` or by setting a new `TargetVersion` every time a new River DB version comes out (not convenient).

I think I'd lean toward documenting this and making sure to have `MaxSteps: 1` in most examples as you suggest.